actors/builtin/miner/miner_actor.go

```diff
@@ -2067,6 +2068,287 @@ func (a Actor) RepayDebt(rt Runtime, _ *abi.EmptyValue) *abi.EmptyValue {
+				newSectorInfo.ReplacedSectorAge = maxEpoch(0, rt.CurrEpoch()-updateWithDetails.sectorInfo.Activation)
+
+				initialPledgeAtUpgrade := InitialPledgeForPower(pwr, rewRet.ThisEpochBaselinePower, rewRet.ThisEpochRewardSmoothed,
+					powRet.QualityAdjPowerSmoothed, rt.TotalFilCircSupply())
+
+				if initialPledgeAtUpgrade.GreaterThan(updateWithDetails.sectorInfo.InitialPledge) {
+					deficit := big.Sub(initialPledgeAtUpgrade, updateWithDetails.sectorInfo.InitialPledge)
+
+					unlockedBalance, err := st.GetUnlockedBalance(rt.CurrentBalance())
+					builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to calculate unlocked balance")
+					builtin.RequirePredicate(rt, unlockedBalance.GreaterThanEqual(deficit), exitcode.ErrInsufficientFunds, "insufficient funds for new initial pledge requirement %s, available: %s, skipping sector %d",
+						deficit, unlockedBalance, updateWithDetails.sectorInfo.SectorNumber)
+
+					err = st.AddInitialPledge(deficit)
+					builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to add initial pledge")
+
+					newSectorInfo.InitialPledge = initialPledgeAtUpgrade
+				}
+
+				var partition Partition
+				found, err := partitions.Get(updateWithDetails.update.Partition, &partition)
+
+				builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadline %v partition %v",
+					updateWithDetails.update.Deadline, updateWithDetails.update.Partition)
+
+				if !found {
+					rt.Abortf(exitcode.ErrNotFound, "no such deadline %v partition %v", dlIdx, updateWithDetails.update.Partition)
+				}
+
+				// TODO: Group by partition
+				partitionPowerDelta, partitionPledgeDelta, err := partition.ReplaceSectors(store,
+					[]*SectorOnChainInfo{updateWithDetails.sectorInfo},
+					[]*SectorOnChainInfo{newSectorInfo},
+					info.SectorSize,
+					quant)
+
+				builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to replace sector at deadline %d partition %d", updateWithDetails.update.Deadline, updateWithDetails.update.Partition)
+
+				powerDelta = powerDelta.Add(partitionPowerDelta)
+				pledgeDelta = big.Add(pledgeDelta, partitionPledgeDelta) // expected to be zero, see note below.
+
+				err = partitions.Set(updateWithDetails.update.Partition, &partition)
+				builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to save deadline %v partition %v",
+					updateWithDetails.update.Deadline,
+					updateWithDetails.update.Partition)
+
+				deadline.Partitions, err = partitions.Root()
+				builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to save partitions for deadline %d", updateWithDetails.update.Deadline)
+
+				err = deadlines.UpdateDeadline(store, updateWithDetails.update.Deadline, deadline)
+				builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to save deadline %d", updateWithDetails.update.Deadline)
```

Comment:
This should move to 2329 after the deadline specific update loop.  It's not wrong but it kills all of the benefits of grouping updates by deadline.  If you do it this way you should just iterate through updates directly.

Reply:
Yes, most definitely